attestation-agent/deps/crypto/src/native/ec.rs

```diff
@@ -0,0 +1,190 @@
+        match wrapping_algorithm {
+            KeyWrapAlgorithm::EcdhEsA256Kw => {
+                let mut deriver = Deriver::new(self.private_key())?;
+                deriver.set_peer(&epk)?;
+                let z = deriver.derive_to_vec()?;
+                let shared_key = concat_kdf(
+                    KeyWrapAlgorithm::EcdhEsA256Kw.as_ref(),
+                    AES_GCM_256_KEY_BITS as usize / 8,
+                    &z,
+                )?;
+                let mut key = vec![0; encrypted_key.len() - 8];
+                let unwrapping_key = AesKey::new_decrypt(&shared_key)
+                    .map_err(|e| anyhow!("failed to create AES unwrapping key: {e:?}"))?;
+                aes::unwrap_key(&unwrapping_key, None, &mut key, &encrypted_key)
+                    .map_err(|e| anyhow!("failed to unwrap key: {e:?}"))?;
+                Ok(key)
+            }
+        }
+    }
+}
+
+fn concat_kdf(alg: &str, target_length: usize, z: &[u8]) -> Result<Vec<u8>> {
+    let target_length_bytes = ((target_length * 8) as u32).to_be_bytes();
+    let alg_len_bytes = (alg.len() as u32).to_be_bytes();
+
+    let mut output = Vec::new();
+    let md = MessageDigest::sha256();
+    let count = target_length.div_ceil(md.size());
+    for i in 0..count {
+        let mut hasher = Hasher::new(md)?;
+        hasher.update(&((i + 1) as u32).to_be_bytes())?;
+        hasher.update(z)?;
+        hasher.update(&alg_len_bytes)?;
+        hasher.update(alg.as_bytes())?;
+        hasher.update(&0_u32.to_be_bytes())?;
+        hasher.update(&0_u32.to_be_bytes())?;
+        hasher.update(&target_length_bytes)?;
+
+        let digest = hasher.finish()?;
+        output.extend(digest.to_vec());
+    }
+
+    if output.len() > target_length {
+        output.truncate(target_length);
+    }
+
+    Ok(output)
+}
+
+#[derive(Clone, Debug)]
+pub struct P256EcKeyPair {
+    private_key: PKey<Private>,
+}
+
+impl P256EcKeyPair {
+    fn private_key(&self) -> &PKey<Private> {
+        &self.private_key
+    }
+
+    pub fn new() -> Result<Self> {
+        let ec_group = EcGroup::from_curve_name(Nid::X9_62_PRIME256V1)?;
+        let ec_key = EcKey::generate(&ec_group)?;
+        let private_key = PKey::from_ec_key(ec_key)?;
+        Ok(Self { private_key })
+    }
+
+    pub fn x(&self) -> Result<Vec<u8>> {
+        let private_key = self.private_key.ec_key().context("must be a ec key")?;
+        let public_key = private_key.public_key();
+        let mut x = BigNum::new()?;
+        let mut _y = BigNum::new()?;
+        let mut ctx = BigNumContext::new()?;
+        public_key.affine_coordinates_gfp(private_key.group(), &mut x, &mut _y, &mut ctx)?;
+        let mut x = x.to_vec();
+        x.resize(32, b'0');
+        Ok(x)
+    }
+
+    pub fn y(&self) -> Result<Vec<u8>> {
+        let private_key = self.private_key.ec_key().context("must be a ec key")?;
+        let public_key = private_key.public_key();
+        let mut _x = BigNum::new()?;
```

Comment:
Would it make more sense to have one function to get both x and y and then maybe have a wrapper around it that just returns if we need it? There is some duplication between these two functions.